src/pipeline.rs

```diff
@@ -1,18 +1,38 @@
 use std::fs::File;
 use std::io::{BufRead, BufReader};
 
-use crate::configuration::Configuration;
+use crate::configuration::{Column, Configuration, FileType};
 use crate::embedding::{calculate_embeddings, calculate_embeddings_mmap};
 use crate::entity::{EntityProcessor, SMALL_VECTOR_SIZE};
 use crate::persistence::embedding::TextFileVectorPersistor;
 use crate::persistence::entity::InMemoryEntityMappingPersistor;
 use crate::persistence::sparse_matrix::InMemorySparseMatrixPersistor;
 use crate::sparse_matrix::{create_sparse_matrices, SparseMatrix};
 use bus::Bus;
-use smallvec::SmallVec;
+use json;
+use smallvec::{smallvec, SmallVec};
 use std::sync::Arc;
 use std::thread;
 
+
+/// Parse a line of JSON and read its columns into a vector for processing.
+fn read_json_columns(line: &str, columns: &Vec<Column>) -> Vec<SmallVec<[String; SMALL_VECTOR_SIZE]>> {
```

Comment:
For JSONs I would prefer to use https://crates.io/crates/serde (most active and popular).

Reply:
Something like that should allow to deserialize to &'str instead for String.

```use serde::Deserialize;
use serde_json;
use smallvec::SmallVec; // enable serde feature

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum ColumnData<'a> {
    Str(&'a str),
    Array(SmallVec<[&'a str; 20]>),
}
```

and later parse with:
```
let value: Result<HashMap<&str, ColumnData>, _> = serde_json::from_str(row);
```

Reply:
I picked the json crate because it's faster than serde (by serde's own benchmarks) and much more lightweight since we don't need most of serde's features.

Reply:
Though, parsing with serde_json to struct seems faster according to: https://github.com/serde-rs/json-benchmark

Also with serde_json one can parse to &str without need of allocating a new strings. I guess this will be faster, though I did not do any specific benchmarks.

We could also try to use: https://github.com/simd-lite/simd-json (maybe behind feature flag or something) which could be even faster!

Reply:
Can't do parsing to struct here though since we don't know in advance what the data will look like. I did try simd-json, but I had problems getting it to compile so I just went with something I could get working for now. It should be easy enough to swap out later regardless.

I believe serde does allocate new strings as well and the string pointer is just from the deserialized value object.